### PR TITLE
Add method signMessage for bitcoin provider

### PR DESCRIPTION
**Pull request by Thiendekaco**

### CODE REVIEW

File: packages/extension-base/src/background/KoniTypes.ts [hunk 1]
```diff
@@ -24,6 +28,7 @@ import { JsonRpcPayload, JsonRpcResponse } from 'web3-core-helpers';
 import { SignerResult } from '@polkadot/types/types/extrinsic';
 import { HexString } from '@polkadot/util/types';
 
+import { BitcoinApiStrategy } from '../../build/services/chain-service/handler/bitcoin/strategy/types';
```
Comment (S2kael) [inline packages/extension-base/src/background/KoniTypes.ts:31]: Update this line, currently use a not existed direction (build folder)

File: packages/extension-base/src/services/request-service/handler/BitcoinRequestHandler.ts [hunk 2]
```diff
@@ -184,46 +204,157 @@ export default class BitcoinRequestHandler {
...
+      }
+    } else {
+      if (t === 'bitcoinSendTransactionRequest') {
+        result.payload = await this.signTransactionBitcoinWithPayload(request as ConfirmationDefinitionsBitcoin['bitcoinSendTransactionRequest'][0], result as ConfirmationDefinitionsBitcoin['bitcoinSendTransactionRequest'][1]);
       }
+    }
 
-      if (t === 'bitcoinSignatureRequest' || t === 'bitcoinSendTransactionRequest') {
-        const isAlwaysRequired = await this.#requestService.settingService.isAlwaysRequired;
+    if (t === 'bitcoinSignatureRequest' || t === 'bitcoinSendTransactionRequest') {
```
Comment (S2kael) [inline packages/extension-base/src/services/request-service/handler/BitcoinRequestHandler.ts:318]: Add `bitcoinSignPsbtRequest` as well

File: packages/extension-base/src/services/request-service/handler/BitcoinRequestHandler.ts [hunk 3]
```diff
@@ -184,46 +204,157 @@ export default class BitcoinRequestHandler {
...
+    }
+
+    const pair = keyring.getPair(account.address);
+
+    // Unlock the pair if it is locked
+    if (pair.isLocked) {
+      keyring.unlockPair(pair.address);
+    }
+
+    const signAtIndexGenerate = signAtIndex ? (isArray(signAtIndex) ? signAtIndex : [signAtIndex]) : [psbt.inputCount];
```
Comment (S2kael) [inline packages/extension-base/src/services/request-service/handler/BitcoinRequestHandler.ts:274]: Why fallback is `psbt.inputCount`

File: packages/extension-base/package.json [hunk 4]
```diff
@@ -89,9 +90,11 @@
     "web3-core-subscriptions": "1.10.0",
     "web3-eth": "1.10.0",
     "web3-eth-contract": "^1.10.0",
-    "web3-utils": "^1.10.0"
+    "web3-utils": "^1.10.0",
+    "yup": "^1.4.0"
```
Comment (S2kael) [inline packages/extension-base/package.json:94]: No use for this library found

File: packages/extension-base/src/background/KoniTypes.ts [hunk 5]
```diff
@@ -1407,9 +1489,10 @@ export interface ConfirmationDefinitions {
 }
 
 export interface ConfirmationDefinitionsBitcoin {
-  bitcoinSignatureRequest: [ConfirmationsQueueItem<BitcoinSignatureRequest>, ConfirmationResult<string>],
+  bitcoinSignatureRequest: [ConfirmationsQueueItem<BitcoinSignatureRequest>, ConfirmationResult<SignMessageBitcoinResult>],
   bitcoinSendTransactionRequest: [ConfirmationsQueueItem<BitcoinSendTransactionRequest>, ConfirmationResult<string>],
-  bitcoinWatchTransactionRequest: [ConfirmationsQueueItem<BitcoinWatchTransactionRequest>, ConfirmationResult<string>]
+  bitcoinWatchTransactionRequest: [ConfirmationsQueueItem<BitcoinWatchTransactionRequest>, ConfirmationResult<string>],
```
Comment (S2kael) [inline packages/extension-base/src/background/KoniTypes.ts:1494]: Can remove this type if need

File: packages/extension-base/src/koni/background/handlers/Extension.ts [hunk 6]
```diff
@@ -972,11 +946,12 @@ export default class KoniExtension {
     return true;
   }
 
+  // deprecated
```
Comment (S2kael) [inline packages/extension-base/src/koni/background/handlers/Extension.ts:949]: Replace with `/** @deprecated */`

File: packages/extension-base/src/koni/background/handlers/State.ts [hunk 7]
```diff
@@ -1177,6 +1182,345 @@ export default class KoniState {
...
+            throw new BitcoinProviderError(BitcoinProviderErrorType.INVALID_PARAMS, t('Not found signature'));
+          }
+        } else {
+          throw new BitcoinProviderError(BitcoinProviderErrorType.USER_REJECTED_REQUEST);
+        }
+      });
+  }
+
+  public async bitcoinSendTransaction (id: string, url: string, networkKey: string, allowedAccounts: string[], transactionParams: BitcoinSendTransactionParams): Promise<string | undefined> {
+    const bitcoinApi = this.getBitcoinApi(networkKey);
```
Comment (S2kael) [inline packages/extension-base/src/koni/background/handlers/State.ts:1321]: In type defined, `networkKey` is `mainnet` or `testnet`, but in apiMap, these keys do not existed

File: packages/extension-base/src/koni/background/handlers/State.ts [hunk 8]
```diff
@@ -1177,6 +1182,345 @@ export default class KoniState {
...
+      throw new EvmProviderError(EvmProviderErrorType.INVALID_PARAMS, t('You have rescinded allowance for this account in wallet'));
+    }
+
+    const pair = keyring.getPair(fromAddress);
+
+    if (!pair) {
+      throw new EvmProviderError(EvmProviderErrorType.INVALID_PARAMS, t('Unable to find account'));
+    }
+
+    if (networkKey === 'mainnet') {
```
Comment (S2kael) [inline packages/extension-base/src/koni/background/handlers/State.ts:1362]: The `networkKey` has been converted to `chainSlug`, so this condition is incorrect

File: packages/extension-base/src/koni/background/handlers/Tabs.ts [hunk 9]
```diff
@@ -1058,27 +1152,193 @@ export default class KoniTabs {
     return await this.#koniState.addTokenConfirm(id, url, tokenInfo);
   }
 
-  public async handle<TMessageType extends MessageTypes> (id: string, type: TMessageType, request: RequestTypes[TMessageType], url: string, port: chrome.runtime.Port): Promise<ResponseTypes[keyof ResponseTypes]> {
-    console.log('handle', type, request, url, port);
+  async bitcoinGetAddresses (url: string, request: RequestArguments): Promise<Response<RequestAddressesResult>> {
+    try {
+      await this.#koniState.authorizeUrlV2(url, {
+        origin: '',
```
Comment (S2kael) [inline packages/extension-base/src/koni/background/handlers/Tabs.ts:1158]: Missing origin

File: packages/extension-base/src/koni/background/handlers/State.ts [hunk 10]
```diff
@@ -1177,6 +1182,345 @@ export default class KoniState {
...
+    const transactionData = { ...transaction };
+    const token = this.chainService.getNativeTokenInfo(networkKey);
+
+    if (eType === ExtrinsicType.TRANSFER_BALANCE) {
+      // @ts-ignore
+      transactionData.tokenSlug = token.slug;
+    }
+
+    // Custom handle this instead of general handler transaction
+    const transactionEmitter = await this.transactionService.addTransaction({
```
Comment (S2kael) [inline packages/extension-base/src/koni/background/handlers/State.ts:1485]: Should split to another handler, shouldn't reuse logic send transaction in extension

File: packages/extension-base/src/background/KoniTypes.ts [hunk 11]
```diff
@@ -1407,9 +1489,10 @@ export interface ConfirmationDefinitions {
 }
 
 export interface ConfirmationDefinitionsBitcoin {
-  bitcoinSignatureRequest: [ConfirmationsQueueItem<BitcoinSignatureRequest>, ConfirmationResult<string>],
+  bitcoinSignatureRequest: [ConfirmationsQueueItem<BitcoinSignatureRequest>, ConfirmationResult<SignMessageBitcoinResult>],
   bitcoinSendTransactionRequest: [ConfirmationsQueueItem<BitcoinSendTransactionRequest>, ConfirmationResult<string>],
```
Comment (S2kael) [inline packages/extension-base/src/background/KoniTypes.ts:1493]: Split a new type for `SendTransfer`

**Comment by saltict** [comment]
🚀 Deployed on https://pr-162--papaya-melba-f5c160.netlify.app

**Pull request merged**